File: frontend/src/api/comments.ts
import { request } from '@/api/request'
import type { CommentItem } from '@/types/comment'

export function fetchParagraphCommentsByChapter(chapterId: number) {
  return request<Record<string, CommentItem[]>>({
    method: 'GET',
    url: `/comments/paragraph/by-chapter/${chapterId}`,
  })
}

export function fetchParagraphComments(chapterId: number, paragraphIndex: number) {
  return request<CommentItem[]>({
    method: 'GET',
    url: '/comments/paragraph',
    params: { chapterId, paragraphIndex },
  })
}

export function createParagraphComment(data: {
  chapterId: number
  paragraphIndex: number
  content: string
  parentId?: number
}) {
  return request<CommentItem>({
    method: 'POST',
    url: '/comments/paragraph',
    data,
  })
}

export function fetchChapterComments(chapterId: number) {
  return request<CommentItem[]>({
    method: 'GET',
    url: `/comments/chapter/${chapterId}`,
  })
}

export function createChapterComment(
  chapterId: number,
  content: string,
  parentId?: number,
) {
  return request<CommentItem>({
    method: 'POST',
    url: '/comments/chapter',
    data: { chapterId, content, parentId },
  })
}

export function deleteComment(commentId: number, type: 'paragraph' | 'chapter' | 'book') {
  return request<null>({
    method: 'DELETE',
    url: `/comments/${commentId}`,
    params: { type },
  })
}

export function pinComment(commentId: number, type: 'paragraph' | 'chapter' | 'book') {
  return request<CommentItem>({
    method: 'POST',
    url: `/comments/${commentId}/pin`,
    params: { type },
  })
}

export function fetchBookCommentsTree(bookId: number) {
  return request<CommentItem[]>({
    method: 'GET',
    url: `/books/${bookId}/comments`,
  })
}

export function createBookComment(bookId: number, content: string, parentId?: number) {
  return request<CommentItem>({
    method: 'POST',
    url: `/books/${bookId}/comments`,
    data: { content, parentId },
  })
}

File: backend/app/api/v1/comments.py
"""评论路由。"""

from fastapi import APIRouter

from app.api.deps import CurrentUser, DbSession
from app.schemas.comment import CreateChapterCommentRequest, CreateParagraphCommentRequest
from app.schemas.common import ok
from app.services import comment_service

router = APIRouter(prefix="/comments", tags=["comments"])


@router.get("/paragraph/by-chapter/{chapter_id}")
def list_paragraph_comments_by_chapter(chapter_id: int, db: DbSession):
    return ok(comment_service.list_paragraph_comments_by_chapter(db, chapter_id))


@router.get("/paragraph")
def list_paragraph_comments(
    db: DbSession,
    chapter_id: int,
    paragraph_index: int,
):
    return ok(comment_service.list_paragraph_comments(db, chapter_id, paragraph_index))


@router.post("/paragraph")
def create_paragraph_comment(
    body: CreateParagraphCommentRequest,
    db: DbSession,
    user: CurrentUser,
):
    return ok(
        comment_service.create_paragraph_comment(
            db,
            body.chapter_id,
            body.paragraph_index,
            user.id,
            body.content,
            body.parent_id,
        ),
    )


@router.get("/chapter/{chapter_id}")
def list_chapter_comments(chapter_id: int, db: DbSession):
    return ok(comment_service.list_chapter_comments(db, chapter_id))


@router.post("/chapter")
def create_chapter_comment(
    body: CreateChapterCommentRequest,
    db: DbSession,
    user: CurrentUser,
):
    return ok(
        comment_service.create_chapter_comment(
            db, body.chapter_id, user.id, body.content, body.parent_id,
        ),
    )


@router.delete("/{comment_id}")
def delete_comment(
    comment_id: int,
    db: DbSession,
    user: CurrentUser,
    type: str,
):
    comment_service.delete_comment(db, user, comment_id, type)
    return ok(None, message="已删除")


@router.post("/{comment_id}/pin")
def pin_comment(
    comment_id: int,
    db: DbSession,
    user: CurrentUser,
    type: str,
):
    return ok(comment_service.pin_comment(db, user, comment_id, type))
